src/main/kotlin/org/rust/ide/refactoring/inlineTypeAlias/RsInlineTypeAliasProcessor.kt

```diff
@@ -0,0 +1,172 @@
+            typeAlias.delete()
+        }
+    }
+
+    private fun partitionUsages(usages: Array<UsageInfo>): Triple<List<PathUsage>, List<RsUseSpeck>, Boolean> {
+        val paths = mutableListOf<PathUsage>()
+        val useSpecks = mutableListOf<RsUseSpeck>()
+        for (usage in usages) {
+            val path = usage.element as? RsPath ?: continue
+            if (path.isExpandedFromMacro) continue
+            val useSpeck = path.ancestorOrSelf<RsUseSpeck>()
+            if (useSpeck != null) {
+                useSpecks += useSpeck
+            } else {
+                val resolved = path.reference?.advancedResolve() ?: continue
+                testAssert { resolved.element == typeAlias }
+                val substitution = tryGetTypeAliasSubstitutionUsingParent(path, typeAlias) ?: resolved.subst
+                paths += PathUsage(path, substitution)
+            }
+        }
+        val hasOtherUsages = paths.size + useSpecks.size != usages.size
+        return Triple(paths, useSpecks, hasOtherUsages)
+    }
+
+    private fun addNeededImports(inlined: List<RsElement>) {
+        val typeReference = typeAlias.typeReference ?: return
+        val handledMods = hashSetOf(typeAlias.containingMod)
+        for (context in inlined) {
+            val mod = context.containingMod
+            if (!handledMods.add(mod)) continue
+            RsImportHelper.importTypeReferencesFromElement(context, typeReference)
+        }
+    }
+
+    override fun getCommandName(): String = "Inline Type Alias ${typeAlias.name}"
+
+    override fun createUsageViewDescriptor(usages: Array<UsageInfo>): UsageViewDescriptor =
+        RsInlineUsageViewDescriptor(typeAlias, "Type Alias to inline")
+}
+
+/**
+ * type Foo<T> = Vec<T>;
+ * fn main() {
+ *     let v = Foo::new();
+ *             ~~~ has substitution `T => T`
+ *             ~~~~~~~~ has substitution `Self => Vec<i32>`
+ *     v.push(1);
+ * }
+ */
+fun tryGetTypeAliasSubstitutionUsingParent(path: RsPath, typeAlias: RsTypeAlias): Substitution? {
+    val parentPath = path.parent as? RsPath ?: return null
+    if (parentPath.parent !is RsPathExpr) return null
+    val resolvedMethod = parentPath.reference?.advancedResolve() ?: return null
+    val selfTy = resolvedMethod.subst[TyTypeParameter.self()] ?: return null
+    val inference = path.implLookup.ctx
+    val subst = inference.instantiateBounds(typeAlias, selfTy)
+    val typeReference = typeAlias.typeReference ?: return null
+    val type = typeReference.normType.substitute(subst)
+    inference.combineTypes(type, selfTy)
+    return subst.mapTypeValues { (_, v) -> inference.resolveTypeVarsIfPossible(v) }
+}
+
+/**
+ * type Foo<T> = Vec<T>;
+ *               ~~~~~~ type
+ * fn main() {
+ *     let _: Foo<i32>;
+ *            ~~~~~~~~ path
+ * }                   subst = "T => i32"
+ */
+fun replacePathWithType(path: RsPath, typeReference0: RsTypeReference, substitution: Substitution): RsElement? {
```

Comment:
Maybe we could choose a more descriptive name for this method, for example, `fillPathWithActualType` (if I understand it right). Also consider renaming `typeReference0` to something like `originalTypeReference`